### ISSUE TITLE
Focus visible: Improve contrast on all backgrounds

### ISSUE THREAD

**Issue by Febakke** [digdir/designsystem#269]
### Describe the component

Today we use a border outside of our component with #98177E as the color. This has high contrast to light background, but might fall short on darker backgrounds. We have solved this with an inverted design today. 
![Skjermbilde 2023-03-28 kl  15 31 49](https://user-images.githubusercontent.com/33222679/228253330-48ff8d6a-bed5-4ed9-9e45-a7ff86da1078.png)
Inverted:
![Skjermbilde 2023-03-28 kl  15 32 21](https://user-images.githubusercontent.com/33222679/228253496-2f9cce53-9ef5-4098-9a59-bf7cef70ddc9.png)

But Gov.uk and others use a combination of yellow and black as their focus state. This ensures high contrast on both dark and light backgrounds. Maybe we should look into solving this in similar way? We are still early in the prosess of making the design system, and having a focus state that works in more scenarios would be easier to work with instead of changing based on the background. 

Example:
![Skjermbilde 2023-03-28 kl  15 34 18](https://user-images.githubusercontent.com/33222679/228254202-77e40fde-306e-42c1-9d07-173f5587efbc.png)

[Gov.uk article on why they solved it like this](https://design-system.service.gov.uk/get-started/focus-states/)
[Also relevant article](https://iknowdavehouse.medium.com/focusing-on-focus-states-7c64089d805c)

**Comment by Camulos** [comment]
Her er kravspec som ble brukt i Brønnøysundregistrene for et par år siden:

- Fokusindikator benytter fargeprofilen
- Fokusindikator oppfyller WCAG krav til fokus
    - https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-visible.html
    - https://www.w3.org/TR/WCAG22/#focus-visible-enhanced (NB! Fjernet/endret i siste utkast til WCAG 2.2)
        - "fysisk" endring som minst har areal som lengden på den lengste siden av elementet som skal ha indikator X 2
            - dvs en komponent som er 48 x 64 px så skal den fysiske endringen være minimum 128px av arealet (64px x 2)
    - Fargekontrast-forskjell mot bakgrunn 7:1 (Minimum 3,5)
    - Fargekontrast-forskjell mot øvrig del av komponent 7:1
    - Fargekontrast-forskjell mot ikke-fokus varianter (dersom alt endres) 4,5:1
- Mørk versjon skal også defineres, støtte preferanse for dark-mode
- Kode: farger skal ha logisk navn i CSS og bygget slik at farger kan endres globalt fra designsystem fremfor i hver enkelt løsning
    - Designsystem betyr: Fokusstil må være definert slik at en kan endre det for alle løsninger fra et sentralt sted
- Valg av fokusstil må fungere på de statiske sidene på brreg.no 
- Valg av fokusstil må fungere i de dynamiske løsningene
- Farge som velges for fremheve fokus skal være forbeholdt kun bruk ved fokus eller andre visuelle fremstillinger for å tydeliggjøre visuelt funksjoner, navigasjon eller liknende.

Sånn ble knapp for å oppfylle disse kravene:
![Skjermbilde 2023-04-26 kl  10 30 26](https://user-images.githubusercontent.com/5771598/234532817-0c03ab86-e4e4-4787-a99e-054a5dd5027f.png)
( https://www.figma.com/file/ydhbA9THirXfz2LbODsBD6/BR-Komponenter%3A-Button?type=design&node-id=2%3A87&t=0a5I0AyCtihH0sYD-1 )
3 px bunnlinje (inside) på hover + endra farge
3 px ramme (inside) på focus + en veldig tydelig farge
For tekstlenker: bakgrunnsfarge + tykkere linje under + fet skrift
For inputfelter og select har vi noe slikt:
![Skjermbilde 2023-04-26 kl  11 35 51](https://user-images.githubusercontent.com/5771598/234535523-e40caaec-dadc-45e7-bdc2-0f4c1efc4717.png)
Altså både at den mørke rammen blir tykkere 1 til 3 px, men at en også legger på en 3 px kontrastfargeramme utenfor.

Dette er endringer I WCAG 2.2 utkast som har kommet senere:
WCAG 2.2 draft 2.4.11 Focus Appearance:

The entire focus indicator meets all the following:
- encloses the user interface component or sub-component that is focused, and
- has a contrast ratio of at least 3:1 between the same pixels in the focused and unfocused states, and
- has a contrast ratio of at least 3:1 against adjacent non-focus-indicator colors.
An area of the focus indicator meets all the following:
- is at least as large as the area of a 1 CSS pixel thick perimeter of the unfocused component or sub-component, or is at least as large as a 4 CSS pixel thick line along the shortest side of the minimum bounding box of the unfocused component or sub-component, and
- has a contrast ratio of at least 3:1 between the same pixels in the focused and unfocused states, and
- has a contrast ratio of at least 3:1 against adjacent non-focus-indicator colors, or is no thinner than 2 CSS pixels.

Siste utkast I WCAG 2.2 gjør at valgene som ble tatt for kontrast hos oss ikke lenger er helt dekkende, fordi det er krav om at endringer skal være utenfor den ytre boksen av elementet.
For å dekke kravet nå så må vi derfor heller gå litt i denne retningen:
![wcag22april2023](https://user-images.githubusercontent.com/5771598/234532873-068b3124-7a0e-417e-97e1-71e62b0a080c.jpg)

Denne versjonen oppfyller kontrastkrav på mer enn minste krav i 2.2-utkastet, også den fysiske endringen er da 3 piksler rundt hele knappen, altså 3 ganger mer enn 2.2-utkastet. 

For andre komponenter, feks lenker så er de valgene vi tok fremdeles nok til å oppfylle alle kravene, uten at vi må endre noe (https://www.figma.com/file/ydhbA9THirXfz2LbODsBD6/BR-Komponenter%3A-Button?type=design&node-id=1524%3A130&t=0a5I0AyCtihH0sYD-1). Her er utfordringen at når en setter fet skrift på hover og focus så vil normalt det føre til at teksten tar mer plass og innhold på en side kan hoppe rundt, det er vanskelig å lage en css-regel som alltid tar høyde for hvor mye mindre tegnavstand en skal ha, men det kan løses med js som holder av plass til det eller regner ut hvor mye mindre tegnsavstand det skal være. Kan være bedre å droppe fet skrift. 
Inputfelter får allerede kontrastramme utenfor og burde være ok:
![Skjermbilde 2023-04-26 kl  11 35 51](https://user-images.githubusercontent.com/5771598/234536594-15381330-d0d2-485b-8ccf-89adc585d5fb.png)

**Comment by mrosvik** [comment]
Vi har kommet frem til at vi i første omgang oppdaterer focus-state til: 
- Box-shadow som gir ekstra bred svart ($border-default / 1E2B3C) border på innsiden (3px)
- Gul outline (3px) #fd0
Dette skal være synlig både ved klikk og tab-focus/tastaturnavigering. 

![image](https://user-images.githubusercontent.com/1464915/235625454-31a5869e-4a7c-4f34-a790-84f267ff381c.png)

Se test av komponenter her: https://uqs7n6.csb.app/

- Vi kan evt. vurdere å endre bakgrunner, men dette vil da kreve mer håndtering av farger for noen komponenter som eksempelvis har elementer av ikoner, sirkler og tekst. (Se f.eks Card B i test-filen https://uqs7n6.csb.app/)
- ToggleGroup vil ikke funke bra med gul bakgrunn, da det ikke blir like tydelig hva som er aktivt. (Se Toggle Button B i test-filen https://uqs7n6.csb.app/)
- Fokus bør i den grad det er mulig være konsekvent på alle interaktive komponenter.

**Comment by Camulos** [comment]
Kan vi i neste runde få en test med tekstlenke som del av et avsnitt, altså når det står som lenke inline.

Toggle komponenten virker som en komponent som kanskje burde være jobbet mer med.

Kan vi også få radioknapp og checkbox inn i testsiden med komponenter?

**Comment by mimarz** [comment]
Me and @TomasEng decided in a meeting on 04.05.23

- We will consolidate focus-styling in one common.modules.css as this is a "global style"
- We discovered some discrepancies when looking at `Radio` & `Checkbox`, focus or how its applied might need to be adjusted here
- We will use `box-shadow` for the black border 3px and yellow 3px on `outline`
- `focus-within` and `focus` selector might be applied
  - Have a look at css-module compose?
